packages/tasit-action/src/TasitAction.js

```diff
@@ -19,117 +22,370 @@ class Utils {
+              new Error(`Your message has been included in an uncle block.`)
+            );
+
+          return;
+        }
+
+        const { confirmations } = receipt;
+        const message = {
+          data: {
+            confirmations: confirmations,
+          },
+        };
+
+        await listener(message);
       } catch (error) {
-        throw new Error(`Callback function with error: ${error.message}`);
+        this.emitErrorEvent(
+          new Error(`Callback function with error: ${error.message}`)
+        );
       }
-    });
+    };
+
+    this.addListener(eventName, "block", listener, wrappedListener);
   };
 
-  removeAllListeners = async () => {
+  // Tech debt
+  // This method avoid duplicated nonce generation when rapid succession of several transactions
+  // See: https://github.com/ethereumbook/ethereumbook/blob/04f66ae45cd9405cce04a088556144be11979699/06transactions.asciidoc#keeping-track-of-nonces
+  // How we'll should keeping track of nonces?
+  waitForMessage = async () => {
     const tx = await this.#txPromise;
-    this.#provider.removeAllListeners(tx.hash);
+
+    await this.#provider.waitForTransaction(tx.hash);
   };
 }
 
-class EventSubscription {
-  #eventNames;
+class ContractSubscription extends Subscription {
+  #contract;
   #provider;
+  #eventNames;
 
-  constructor(eventNames, provider) {
-    this.#eventNames = eventNames;
+  // Why eventNames shoud be passed here?
+  constructor(contract, provider, eventNames) {
+    eventNames.forEach(eventName => {
+      if (contract.interface.events[eventName] === undefined)
+        throw new Error(`Event '${eventName}' not found.`);
+    });
+
+    super(contract);
+    this.#contract = contract;
     this.#provider = provider;
+    this.#eventNames = eventNames;
   }
 
-  on = (eventName, callback) => {
-    if (!this.#eventNames.includes(eventName))
+  on = (eventName, listener) => {
+    if (!this.#eventNames.includes(eventName) && eventName !== "error") {
       throw new Error(
         `This subscription isn't subscribed on '${eventName}' event.`
       );
+    }
+
+    if (eventName === "error") {
+      this.addErrorListener(listener);
+      return;
+    }
+
+    const wrappedListener = async (...args) => {
+      const event = args.pop();
+      const message = {
+        data: {
+          args: event.args,
+        },
+      };
+
+      try {
+        await listener(message);
+      } catch (error) {
+        this.emitErrorEvent(
+          new Error(`Callback function with error: ${error.message}`)
+        );
+      }
+    };
+
+    this.addListener(eventName, eventName, listener, wrappedListener);
   };
+}
 
-  //removeAllListeners = async () => {};
+class ProviderFactory {
+  static getProvider = () => {
+    let json;
+    try {
+      json = PROVIDER_CONFIG;
+    } catch (error) {
+      console.warn(
+        `Error on parsing config.json file, using default configuration.`
+      );
+      json = ProviderFactory.getDefaultConfig();
```

Comment:
It's a deprecated try/catch, this error was called because the code was import the file directly using require('config.json').
Now, we're using `config` module. I'll see how that's handle with this kind of error.
How should be our behavior in these situations:
1. config file not found (Throw error or use default config?)
2. config w/ missing parts (e.g. jsonrpc provider without connections params) (Throw error or use default config?)
3. error on parsing config file (Throw error and exit or warn use default config?)

Reply:
1. I'm leaning towards throw error but keep an option open for the user to use the default as simply as possible.
2. I'm leaning towards throw error and log exactly what is missing and suggest defaults in the log statements. The more lines we have `const { bla } = blaBla` followed by
```javascript
if (!bla) {
  console.log("bla is needed to do foo because bar, but it's missing")
}
```
the better.
3. I'm leaning towards throwing and error and logging as best we can in which step the parsing failed. The more lines we have like this, the better:
```javascript
try {
  jsonObject = JSON.parse(jsonString)
} catch (error) {
  if (error.message === "something") // or more likely, whether it contains a word
    console.log("This is why it failed")
  }
}
```

Reply:
Thread extract to issue: https://github.com/tasitlabs/TasitSDK/issues/103

Reply:
I'm resolving this since it's extracted